### PR TITLE
Share recurrent state under CUDA graph capture

### PR DESCRIPTION
**Pull request by tianleiwu**
## Summary

Allow recurrent models to use one state buffer under CUDA graph capture. The first capture backs up and restores the committed state so capture-time writes do not leak into generation; subsequent graph replay can safely update the shared buffer in place. 

It also fixed the multimodel cuda graph issue mentioned in https://github.com/microsoft/onnxruntime-genai/pull/2446.

## Key changes

- Detect the first recurrent CUDA graph capture in `DecoderOnly_State::Run`.
- Save committed recurrent and convolution state before capture and restore it afterward.
- Use one shared recurrent-state allocation by default, halving this state storage and avoiding a second graph-buffer variant.
- Retain double buffering as a diagnostic fallback with `ORTGENAI_SHARE_RECURRENT_STATE_UNDER_GRAPH_CAPTURE=0`.

## Performance and memory

For the 48-layer Qwen3.8 GatedDeltaNet model with `state_window=4`, one state set is 587.25 MiB. Three 4K-token N=3 runs averaged 209.90 tok/s with one buffer and produced identical hashes, target-forward counts, and acceptance.

## Validation

- Built `python` and `unit_tests` with CUDA.
- Focused MTP/speculative suite: 51 passed.
- `lintrunner` and `git diff --check`: clean.
- Graph-stable reset/replay smoke passed with the single-buffer default.

### CODE REVIEW

File: src/models/recurrent_state.cpp [hunk 1]
```diff
@@ -118,15 +118,15 @@ RecurrentState::RecurrentState(State& state)
...
+  // Under CUDA-graph capture the recurrent (conv + linear-attention) state is updated in
+  // place (present_state aliased onto past_state), and ORT re-runs the model several times
+  // inside the first Run() of each captured shape to warm up and capture. Save/restore around
+  // that first capture (see ShouldFixUpGraphCapture) keeps the update correct while using half
+  // the recurrent-state memory and one graph variant. The environment override can disable
+  // sharing to retain double buffering as a diagnostic fallback.
+  bool share_under_graph_capture = true;
+  GetEnv("ORTGENAI_SHARE_RECURRENT_STATE_UNDER_GRAPH_CAPTURE", share_under_graph_capture);
+  graph_double_buffer_ = !is_webgpu && state_.params_->use_graph_capture && !share_under_graph_capture;
   share_buffers_ = !is_webgpu && !graph_double_buffer_;
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/models/recurrent_state.cpp]: Fixed in `bef1f498`. Non-WebGPU runs now preserve the configured `past_present_share_buffer` setting outside graph capture. Graph capture only forces sharing when the capture-specific override permits it; otherwise it uses the diagnostic double-buffered path.

File: src/models/recurrent_state.cpp [hunk 2]
```diff
@@ -331,6 +331,32 @@ void RecurrentState::RestoreSnapshot() {
...
+    span.CopyDeviceToCpu();
+    graph_capture_backup_.push_back(std::move(span));
+  }
+}
+
+void RecurrentState::RestoreAfterGraphCapture(int graph_id) {
+  for (auto& span : graph_capture_backup_) {
+    span.CopyCpuToDevice();
+  }
+  graph_capture_backup_.clear();
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/models/recurrent_state.cpp:361]: Fixed in `bef1f498`. The restore path now synchronizes the recurrent-state device stream after the asynchronous host-to-device copies and before clearing the backup spans, so their pinned CPU storage remains alive until the copies complete.

File: src/models/recurrent_state.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tianleiwu) [reply src/models/recurrent_state.cpp]: Fixed in `bef1f498`. The multimodal decoder now uses the same save/capture/restore/replay flow and graph-variant keying as the decoder-only path. I also extended the CUDA fixture with observable recurrent state and added an eager-vs-graph logits regression test.

File: src/models/recurrent_state.cpp [hunk 3]
```diff
@@ -331,6 +334,34 @@ void RecurrentState::RestoreSnapshot() {
...
 }
 
+bool RecurrentState::ShouldFixUpGraphCapture(int graph_id) const {
+  if (layer_indices_.empty() || !share_buffers_ || !state_.params_->use_graph_capture)
+    return false;
+  return std::find(graph_capture_fixed_up_.begin(), graph_capture_fixed_up_.end(), graph_id) ==
+         graph_capture_fixed_up_.end();
+}
+
+void RecurrentState::SaveForGraphCapture() {
```
Comment (baijumeswani) [inline src/models/recurrent_state.cpp:344]: optimization follow-up:

> This performs one pinned-host allocation, device-to-host copy, and stream synchronization per recurrent tensor. A 48-layer model means 96 serialized copies/synchronizations and roughly 587 MiB of temporary pinned host memory. This only occurs once per captured shape, so steady-state throughput improves, but TTFT and host-memory pressure should be measured; a contiguous/reusable staging buffer would reduce overhead.

**Review by baijumeswani — APPROVED** [review]

File: src/models/recurrent_state.cpp (hunk 3 above)
Comment (tianleiwu) [reply src/models/recurrent_state.cpp:344]: This is necessary to avoid double buffer: we can use either device memory or host memory to back up the input state in cuda graph capture, I chose host memory because it reduces device memory usage.

cuda graph capture is one only once for per captured shape, and impact is small.

Here is related issue: https://github.com/microsoft/onnxruntime/issues/32243
